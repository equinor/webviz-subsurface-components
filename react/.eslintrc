{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "extends":  [
    "plugin:react/recommended",
    "eslint:recommended",
    "plugin:prettier/recommended"
  ],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"],
      "plugins": ["@typescript-eslint"],
      "extends": [
        "plugin:react/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:prettier/recommended"
      ]
    }
  ],
  "settings": {
    "react": {
      "version": "detect"
    }
  },
  "env": {
    "amd": true,
    "browser": true,
    "es6": true,
    "jasmine": true,
    "node": true
  }
}
