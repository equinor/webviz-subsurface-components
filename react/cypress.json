{
    "baseUrl": "http://localhost:6006/?path=/story/deckglmap--default",
    "video": true
}
