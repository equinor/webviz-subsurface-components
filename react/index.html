<html>
    <head>
        <title>Demo application of webviz-subsurface-components</title>
    </head>
    <body>
        <div id='root'></div>
        <link rel="stylesheet" type="text/css" href="output.css">
        <script src="output.js"></script>
    </body>
</html>
