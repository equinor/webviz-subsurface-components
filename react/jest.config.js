/**
 * The jest tests had an error: SyntaxError: Cannot use import statement outside a module
 * After some investigation, it leads to this page https://github.com/facebook/jest/issues/9395.
 */
module.exports = {
    moduleNameMapper: {
        "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$":
            "<rootDir>/__mocks__/fileMock.js",
        "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js",
    },
    transform: {
        "\\.(js|ts|jsx|tsx)$": [
            "babel-jest",
            { configFile: "./config/babel.config.json" },
        ],
    },
    transformIgnorePatterns: ["<rootDir>/node_modules/(?!(@webviz)/)"],
};
