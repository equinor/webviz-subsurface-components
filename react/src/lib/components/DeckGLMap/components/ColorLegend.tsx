import React from "react";
import { Layer } from "deck.gl";
import {
    DiscreteColorLegend,
    ContinuousLegend,
} from "@emerson-eps/color-tables";
import { colorTablesArray } from "@emerson-eps/color-tables/";

interface ColorLegendProps {
    position?: number[] | null;
    horizontal?: boolean | null;
    layers: Layer<unknown>[];
    colorTables: colorTablesArray;
}

// Todo: Adapt it for other layers too
const ColorLegend: React.FC<ColorLegendProps> = ({
    position,
    horizontal,
    layers,
}: ColorLegendProps) => {
    const [legendProps, setLegendProps] = React.useState<
        [
            {
                title: string;
                colorName: string;
                discrete: boolean;
                metadata: { objects: Record<string, [number[], number]> };
                valueRange: number[];
                visible: boolean;
            }
        ]
    >([
        {
            title: "",
            colorName: "string",
            discrete: false,
            metadata: { objects: {} },
            valueRange: [],
            visible: true,
        },
    ]);

    // Get color table for log curves.
    React.useEffect(() => {
        if (!layers) return;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const getLegendData: any = [];
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        layers.map((layer: any) => {
            if (
                layer?.id == "wells-layer" &&
                layer?.isLoaded &&
                Object.keys(layer?.state).length > 0
            ) {
                getLegendData.push({
                    title: layer?.state?.legend[0].title,
                    colorName: layer?.props?.logColor,
                    discrete: layer?.state?.legend[0].discrete,
                    metadata: layer?.state?.legend[0].metadata,
                    valueRange: layer?.state?.legend[0].valueRange,
                    visible: layer?.props?.visible,
                });
            }
            if (
                layer?.id == "colormap-layer" &&
                layer?.isLoaded &&
                Object.keys(layer?.state).length > 0
            ) {
                const min = layer?.state.model.uniforms.colorMapRangeMin;
                const max = layer?.state.model.uniforms.colorMapRangeMax;
                getLegendData.push({
                    title: layer?.props?.name,
                    colorName: layer?.props?.colorMapName,
                    discrete: false,
                    metadata: { objects: {} },
                    valueRange: [min, max],
                    visible: layer?.props?.visible,
                });
            }
        });
        setLegendProps(getLegendData);
    }, [layers]);

    return (
        <div
            style={{
                position: "absolute",
                display: "flex",
                right: position ? position[0] : " ",
                top: position ? position[1] : " ",
                zIndex: 999,
            }}
        >
            {legendProps.map(
                (legend, index) =>
                    legend.visible && (
                        <div key={index}>
                            {legend.discrete && (
                                <DiscreteColorLegend
                                    discreteData={legend.metadata}
                                    dataObjectName={legend.title}
                                    colorName={legend.colorName}
                                    position={position}
                                    horizontal={horizontal}
                                />
                            )}
                            {legend.valueRange?.length > 0 && legend && (
                                <ContinuousLegend
                                    min={legend.valueRange[0]}
                                    max={legend.valueRange[1]}
                                    dataObjectName={legend.title}
                                    colorName={legend.colorName}
                                    position={position}
                                    horizontal={horizontal}
                                />
                            )}
                        </div>
                    )
            )}
        </div>
    );
};

ColorLegend.defaultProps = {
    position: [5, 10],
    horizontal: false,
};

export default ColorLegend;
