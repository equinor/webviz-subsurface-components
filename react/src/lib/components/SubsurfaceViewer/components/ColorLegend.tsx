import React from "react";
import {
    DiscreteColorLegend,
    ContinuousLegend,
} from "@emerson-eps/color-tables";
import { ExtendedLayer } from "../layers/utils/layerTools";
import { Color } from "@deck.gl/core/typed";
import { colorTablesArray } from "@emerson-eps/color-tables/";
import { colorMapFunctionType } from "../layers/utils/layerTools";

interface LegendBaseData {
    title: string;
    colorName: string;
    discrete: boolean;
    colorMapFunction?: colorMapFunctionType;
}
export interface DiscreteLegendDataType extends LegendBaseData {
    metadata: Record<string, [Color, number]>;
}

export interface ContinuousLegendDataType extends LegendBaseData {
    valueRange: [number, number];
}

interface ColorLegendProps {
    horizontal?: boolean | null;
    layer: ExtendedLayer<unknown>;
    colorTables: colorTablesArray | string | undefined;
    reverseRange?: boolean;
}

const ColorLegend: React.FC<ColorLegendProps> = ({
    horizontal,
    layer,
    colorTables,
    reverseRange,
}: ColorLegendProps) => {
    const [legendData, setLegendData] = React.useState<
        DiscreteLegendDataType | ContinuousLegendDataType
    >();
    React.useEffect(() => {
        const legend_data = layer.getLegendData?.() ?? layer.state?.["legend"];
        setLegendData(legend_data);
    }, [layer.props, layer.state?.["legend"]]);

    if (!legendData || !layer.props.visible) return null;

    return (
        <div style={{ marginTop: 30 }}>
            {legendData.discrete && (
                <DiscreteColorLegend
                    discreteData={
                        (legendData as DiscreteLegendDataType).metadata
                    }
                    dataObjectName={legendData.title}
                    colorName={legendData.colorName}
                    horizontal={horizontal}
                    colorTables={colorTables}
                />
            )}
            {!legendData.discrete && (
                <ContinuousLegend
                    min={(legendData as ContinuousLegendDataType).valueRange[0]}
                    max={(legendData as ContinuousLegendDataType).valueRange[1]}
                    dataObjectName={legendData.title}
                    colorName={legendData.colorName}
                    horizontal={horizontal}
                    id={layer.props.id}
                    colorTables={colorTables}
                    colorMapFunction={legendData.colorMapFunction}
                    reverseRange={reverseRange}
                />
            )}
        </div>
    );
};

export default ColorLegend;
