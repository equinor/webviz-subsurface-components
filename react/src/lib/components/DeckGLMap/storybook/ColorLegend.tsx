import * as React from "react";
import { ContinuousLegend } from "./ContinuousLegend";
import { DiscreteColorLegend } from "./DiscreteLegend";
import { useCallback } from "react";
import { ColorSelectorAccordion } from "./ColorSelectorAccordion";
import { d3ColorScales } from "./d3ColorScale";
import * as d3 from "d3";

declare type ColorLegendProps = {
    colorTables: any;
    min: number;
    max: number;
    dataObjectName: string;
    position?: number[] | null;
    colorName: string;
    horizontal?: boolean | null;
    discreteData: { objects: Record<string, [number[], number]> };
    getColorMapname?: any;
};

// Todo: Adapt it for other layers too
const ColorLegend: React.FC<ColorLegendProps> = ({
    position,
    horizontal,
    colorTables,
    min,
    max,
    dataObjectName,
    colorName,
    discreteData,
    getColorMapname,
}: ColorLegendProps) => {
    const [isToggled, setIsToggled] = React.useState(false);
    const handleClick = useCallback(() => {
        setIsToggled(true);
    }, []);

    const [updateLegend, setUpdateLegendColor] = React.useState([] as any);
    const [isCont, setIsCont] = React.useState(true);

    // Get new colorscale from colorselector and update legend
    const colorScaleObject = React.useCallback((data: any, value: any) => {
        // colortable colors name
        if (data.name && getColorMapname) {
            getColorMapname(data.name);
        }
        // d3 color name
        else if (getColorMapname) {
            getColorMapname(data.legendColorName);
            //getColorMapname(colorMapping);
        }
        setUpdateLegendColor(data);
        setIsCont(value);
    }, []);

    return (
        <div>
            <div onClick={handleClick}>
                {isCont && (
                    <ContinuousLegend
                        min={min}
                        max={max}
                        dataObjectName={dataObjectName}
                        position={position}
                        colorName={colorName}
                        colorTables={colorTables}
                        horizontal={horizontal}
                        updateLegend={updateLegend}
                    />
                )}
                {!isCont && (
                    <DiscreteColorLegend
                        discreteData={discreteData}
                        dataObjectName={dataObjectName}
                        position={position}
                        colorName={colorName}
                        colorTables={colorTables}
                        horizontal={horizontal}
                        updateLegend={updateLegend}
                    />
                )}
            </div>
            <div>
                {isToggled && (
                    <ColorSelectorAccordion
                        colorScaleObject={colorScaleObject}
                        isHorizontal={horizontal}
                    />
                )}
            </div>
        </div>
    );
};

export default ColorLegend;
