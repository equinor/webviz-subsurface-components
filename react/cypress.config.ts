import { defineConfig } from "cypress";
import getCompareSnapshotsPlugin from "cypress-image-diff-js/dist/plugin";

export default defineConfig({
  component: {
    devServer: {
      framework: "create-react-app",
      bundler: "webpack",
    },
    video:false,
    setupNodeEvents(on, config) {
      getCompareSnapshotsPlugin(on, config),
      on('before:browser:launch', (browser = {}, launchOptions) => {
        if (browser.family === 'chromium' && browser.name !== 'electron') {
          launchOptions.args.push('--start-fullscreen')
      
          return launchOptions
        }
        if (browser.name === 'electron') {
          launchOptions.preferences.fullscreen = true
      
          return launchOptions
        }
      })
    }
  },
});
