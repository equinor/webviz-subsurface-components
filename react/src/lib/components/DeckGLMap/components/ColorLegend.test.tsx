/* eslint-disable @typescript-eslint/no-var-requires */
import { render } from "@testing-library/react";
import "jest-styled-components";
import "@testing-library/jest-dom";
import React from "react";
import ColorLegend from "./ColorLegend";
import { Layer } from "deck.gl";
import {
    getLayersInViewport,
    getLayersWithDefaultProps,
} from "../layers/utils/layerTools";

const example_color_table = require("../../../../demo/example-data/color-tables.json");
const exampleData = require("../../../../demo/example-data/deckgl-map.json");
const layers = getLayersWithDefaultProps(exampleData[0].layers);

describe("Test Color Legend", () => {
    it("snapshot test", () => {
        const { container } = render(
            <ColorLegend
                layers={
                    getLayersInViewport(layers, [
                        "geojson-line-layer",
                        "geojson-layer",
                        "text-layer",
                    ]) as Layer<unknown>[]
                }
                colorTables={[example_color_table]}
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
