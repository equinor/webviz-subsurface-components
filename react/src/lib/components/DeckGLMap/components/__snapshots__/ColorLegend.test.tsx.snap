// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Color Legend snapshot test 1`] = `
<div
  style="position: absolute; display: flex; right: 5px; top: 10px; z-index: 999;"
/>
`;
